thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c

```diff
@@ -1043,6 +1054,174 @@ ngx_field_from_zmauthtoken(ngx_log_t *log, ngx_pool_t *pool,
     return f;
 }
 
+/* extract the jwt token from Header*/
+static ngx_flag_t
+ngx_get_authheader_bearer(ngx_log_t *log, ngx_pool_t *pool,
+        ngx_http_headers_in_t *headers_in, ngx_str_t *value)
+{
+    ngx_log_debug0 (NGX_LOG_DEBUG_HTTP, log, 0,
+            "zmauth: ngx_get_authheader_bearer entered");
+
+    ngx_str_t authValue;
+    ngx_str_t token;
+
+    authValue = headers_in->authorization->value;
+
+    ngx_log_debug1 (NGX_LOG_DEBUG_HTTP, log, 0, "zmauth: ngx_get_authheader_bearer token:%V",&authValue);
+
+    if (authValue.len >= sizeof("Bearer ") - 1
+            && ngx_memcmp(authValue.data,"Bearer ",sizeof("Bearer ")-1)
+            == 0)
+    {
+        ngx_log_debug0 (NGX_LOG_DEBUG_HTTP, log, 0, "zmauth: ngx_get_authheader_bearer Bearer token found");
+
+        token = authValue;
+        token.data += (sizeof("Bearer ") - 1);
+        token.len -= (sizeof("Bearer ") - 1);
+
+        ngx_log_debug1 (NGX_LOG_DEBUG_HTTP, log, 0, "zmauth: ngx_get_authheader_bearer Bearer token found:%V",&token);
+        *value = token;
+
+        return 1; 
+    }
+
+    ngx_log_debug0 (NGX_LOG_DEBUG_HTTP, log, 0,
+            "zmauth: ngx_get_authheader_bearer exit");
+
+    return 0;
+}
+
+/* extract claim data from jwt token */
+static ngx_flag_t
+ngx_claimdata_from_jwttoken(ngx_log_t *log, ngx_pool_t *pool,
+        ngx_str_t *authtoken, ngx_str_t *field, ngx_str_t *value) {
+
+    ngx_flag_t f;
+    ngx_log_debug0 (NGX_LOG_DEBUG_HTTP, log, 0,
+            "zmauth: ngx_claimdata_from_jwttoken entered");
+
+    ngx_str_t pattern = ngx_string("(?<=\\.)(.*?)(?=\\.)");
+    ngx_regex_compile_t rgc;
+    u_char errstr[NGX_MAX_CONF_ERRSTR];
+    ngx_memzero(&rgc, sizeof(ngx_regex_compile_t));
+
+    rgc.pattern  = pattern;
+    rgc.pool     = pool;
+    rgc.err.len  = NGX_MAX_CONF_ERRSTR;
+    rgc.err.data = errstr;
+
+
+    if (ngx_regex_compile(&rgc) != NGX_OK)
+    {
+        ngx_log_error(NGX_LOG_INFO, log, 0,"zmauth: ngx_claimdata_from_jwttoken regex compilation failed.Error=====%s",errstr);
+        return 0;
+    }
+
+    int captures[30];
```

Comment:
Looking at http://nginx.org/en/docs/dev/development_guide.html it appears the captures value should derive from ```rgc```? Is there a reason we're using 30 here? Based on my quick read, that value doesn't seem to correspond with the number of captures we're expecting (nor using). Am I missing something?

Reply:
We can derive it from rgc.